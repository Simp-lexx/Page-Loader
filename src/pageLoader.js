import { promises as fsPromises } from 'fs';
import axios from 'axios';
import path from 'path';
import url from 'url';
import debug from 'debug';
import {
  formatName, buildFilesArray, getDirectLinks, updateLocalLinks,
} from './utils';

const writeLog = debug('page-loader');

export default (sourceUrl, destPath) => {
  const { protocol, hostname } = url.parse(sourceUrl);
  if (!protocol && !hostname) {
    const message = `Incorrent url ${sourceUrl}`;
    throw new Error(message);
  }

  const destFileName = formatName(sourceUrl, '.html');
  const destDirName = formatName(sourceUrl, '_files');

  const destFullFilePath = path.join(destPath, destFileName);
  const destDirPath = path.join(destPath, destDirName);

  let filesLinks = [];

  return axios.get(sourceUrl)
    .then((response) => {
      const getFilesArray = buildFilesArray(response.data);
      filesLinks = getDirectLinks(getFilesArray, sourceUrl);
      const htmlLocalLinks = updateLocalLinks(response.data, destDirName);
      writeLog(`Source page: ${sourceUrl}`);
      writeLog(`Destination file path: ${destFullFilePath}`);
      return fsPromises.writeFile(destFullFilePath, htmlLocalLinks);
    }).catch(error => console.error(error))
    .then(() => fsPromises.mkdir(destDirPath))
    .catch(error => console.error(error))
    .then(writeLog(`Dir is created ${destDirPath}`))
    .then(() => Promise.all(filesLinks.map(link => axios
      .get(link, { responseType: 'arraybuffer' }).catch(error => console.error(error))
      .then((response) => {
        const { pathname } = url.parse(link);
        const currentPathName = formatName(pathname.substring(1));
        const localFullPath = path.join(destDirPath, currentPathName);
        writeLog(`Write downloaded data to: ${localFullPath}`);
        return fsPromises.writeFile(localFullPath, response.data);
      }))))
    .catch((error) => {
      console.log(`Error ${error.message} occured.`);
      console.error(error);
      return Promise.reject(error);
    });
};
