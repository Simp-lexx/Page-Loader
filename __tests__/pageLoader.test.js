import os from 'os';
import path from 'path';
import url from 'url';
import { promises as fsPromises } from 'fs';
import nock from 'nock';
import axios from 'axios';
import httpAdapter from 'axios/lib/adapters/http';
import pageLoader from '../src';

axios.defaults.adapter = httpAdapter;

test('Test download page to local folder', async () => {
  const host = 'https://hexlet.io';
  const pathname = '/courses';
  const expected = '<html><head></head><body></body></html>';

  const sourceFilePath = url.format({ host, pathname });
  const tmpDirName = path.resolve(os.tmpdir(), 'page-loader-');
  const tmpDir = await fsPromises.mkdtemp(tmpDirName);

  nock(host).get(pathname).reply(200, expected);

  await pageLoader(sourceFilePath, tmpDir);
  const destFileName = await fsPromises.readdir(tmpDir);
  const destFilePath = path.join(tmpDir, destFileName[0]);
  const received = await fsPromises.readFile(destFilePath, 'utf8');

  expect(received).toBe(expected);
});

test('Test download page with links', async () => {
  const host = 'https://hexlet.io';
  const pathname = '/';
  const imgPath = '/test/image.jpg';
  const scriptPath = '/test/script.js';
  const linkPath = '/test/style.css';
  const filesBody = 'body';
  const testBody = '<html><head></head><body><img src="test/image.jpg"><script src="test/script.js"></script><link href="test/style.css"></body></html>';
  const expectedBody = '<html><head></head><body><img src="hexlet-io_files/test-image.jpg"><script src="hexlet-io_files/test-script.js"></script><link href="hexlet-io_files/test-style.css"></body></html>';

  const sourceFilePath = url.format({ host, pathname });
  const tmpDirName = path.resolve(os.tmpdir(), 'page-loader-');
  const tmpDir = await fsPromises.mkdtemp(tmpDirName);

  nock(host)
    .get(pathname)
    .reply(200, testBody)
    .get(imgPath)
    .reply(200, filesBody)
    .get(scriptPath)
    .reply(200, filesBody)
    .get(linkPath)
    .reply(200, filesBody);

  await pageLoader(sourceFilePath, tmpDir);
  const destFileName = await fsPromises.readdir(tmpDir);
  const destFilePath = path.join(tmpDir, destFileName[0]);
  const received = await fsPromises.readFile(destFilePath, 'utf8');

  expect(received).toBe(expectedBody);
});

describe('Error tests', () => {
  const host = 'http://test.com';
  test('wrong output', async () => {
    const testPath = '/';
    const link = host.concat(testPath);
    const output = 'folder/folder';
    try {
      await pageLoader(link, output);
    } catch (e) {
      expect(e).toThrowErrorMatchingSnapshot();
    }
  });
  test('incorrect url', async () => {
    const tempDir = await fsPromises.mkdtemp(
      path.resolve(os.tmpdir(), 'page-loader-'),
    );
    const link = 'incorrect url';
    const output = tempDir;
    try {
      await pageLoader(link, output);
    } catch (e) {
      expect(e).toThrowErrorMatchingSnapshot();
    }
  });
  test('can\'t load resource', async () => {
    const document = '<!DOCTYPE html><html><head><link href="/folder/file.css"></head><body></body></html>';
    const tempDir = await fsPromises.mkdtemp(
      path.resolve(os.tmpdir(), 'page-loader-'),
    );
    const testPath = '/';
    const link = host.concat(testPath);
    const output = tempDir;

    nock(host)
      .get(testPath)
      .reply(200, document);

    try {
      await pageLoader(link, output);
    } catch (e) {
      expect(e).toThrowErrorMatchingSnapshot();
    }
  });
});
