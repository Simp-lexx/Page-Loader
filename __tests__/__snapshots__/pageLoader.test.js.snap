// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error tests can't load resource 1`] = `"received is not a function"`;

exports[`Error tests incorrect url 1`] = `"received is not a function"`;
